.github/workflows/ci.yaml

```diff
@@ -0,0 +1,549 @@
+          python -m pip install -U pip setuptools wheel
+          pip uninstall -y astroid
+          pip install -U -r requirements_test.txt
+          pip install -U -r requirements_docs.txt
+          pip install -e .
+      - name: Generate pre-commit restore key
+        id: generate-pre-commit-key
+        run: >-
+          echo "::set-output name=key::${{ runner.os }}-pre-commit-${{
+            env.CACHE_VERSION }}-${{ hashFiles('.pre-commit-config.yaml') }}"
+      - name: Restore pre-commit environment
+        id: cache-precommit
+        uses: actions/cache@v2
+        with:
+          path: ${{ env.PRE_COMMIT_CACHE }}
+          key: ${{ steps.generate-pre-commit-key.outputs.key }}
+          restore-keys: |
+            ${{ runner.os }}-pre-commit-${{ env.CACHE_VERSION }}-
+      - name: Install pre-commit dependencies
+        if: steps.cache-precommit.outputs.cache-hit != 'true'
+        run: |
+          . venv/bin/activate
+          pre-commit install --install-hooks
+
+  pylint:
+    name: Run pylint checks
+    runs-on: ubuntu-latest
+    needs: prepare-base
+    steps:
+      - name: Check out code from GitHub
+        uses: actions/checkout@v2
+      - name: Set up Python ${{ env.DEFAULT_PYTHON }}
+        id: python
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ env.DEFAULT_PYTHON }}
+      - name: Restore Python virtual environment
+        id: cache-venv
+        uses: actions/cache@v2
+        with:
+          path: venv
+          key: >-
+            ${{ runner.os }}-base-venv-${{ steps.python.outputs.python-version }}-${{
+            needs.prepare-base.outputs.python-key }}
+      - name: Fail job if Python cache restore failed
+        if: steps.cache-venv.outputs.cache-hit != 'true'
+        run: |
+          echo "Failed to restore Python venv from cache"
+          exit 1
+      - name: Run pylint check
+        run: |
+          . venv/bin/activate
+          pip install -e .
+          pylint -rn --rcfile=pylintrc \
+            --load-plugins=pylint.extensions.docparams, pylint.extensions.mccabe \
+            pylint \
+            tests/message/ \
+            tests/checkers/ \
+            tests/extensions/ \
+            tests/utils/ \
+            tests/acceptance/ \
+            tests/conftest.py \
+            tests/test_config.py \
+            tests/test_func.py \
+            tests/test_functional.py \
+            tests/test_import_graph.py \
+            tests/test_pragma_parser.py \
+            tests/test_pylint_runners.py \
+            tests/test_regr.py \
+            tests/test_self.py \
+            tests/unittest_config.py \
+            tests/lint/ \
+            tests/unittest_pyreverse_diadefs.py \
+            tests/unittest_pyreverse_inspector.py \
+            tests/unittest_pyreverse_writer.py \
+            tests/unittest_reporters_json.py \
+            tests/unittest_reporting.py
+
+  formatting:
+    name: Run formatting checks
```

Comment:
I think we can remove the pylint and mypy job and rename this. Pre-commit check formatting mypy and pylint. Maybe we need to install pylint first as it need to be system installer for pre-commit.

Reply:
Sounds good. I do install pylint before running pre-commit, although technically the dependencies are already installed after the first job. `pip install -e .` is just necessary since symlinks aren't cached. Just `pip install .` wouldn't work for caching, since we would need to recreate it on every change which kind of defeats the purpose.
https://github.com/PyCQA/pylint/blob/56191d2b849693110c011d850b83a48d797483d2/.github/workflows/ci.yaml#L176-L178